File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/iterator/concrete_aggregate.dart
import 'aggregate.dart';
import 'types.dart';
import 'concrete_iterator.dart';

class Playlist implements Aggregate<Song> {
  final List<Song> _items = [];

  void add(Song s) => _items.add(s);

  int get length => _items.length;

  Song getAt(int i) => _items[i];

  void removeAt(int i) => _items.removeAt(i);

  @override
  PlaylistIterator createIterator() => PlaylistIterator(this);
}

File: implementations/react/src/patterns/behavioral/iterator/concrete-aggregate.ts
import type { Aggregate } from "./aggregate";
import { PlaylistIterator } from "./concrete-iterator";
import type { Song } from "./types/type";
import type { Iterator } from "./iterator";

/**
 * Agregado concreto que representa una lista de reproducción de canciones.
 */

export class Playlist implements Aggregate<Song> {
  private items: Song[] = [];
  add(song: Song) { this.items.push(song); }
  get length() { return this.items.length; }
  getAt(i: number) { return this.items[i]; }
  removeAt(i: number) { this.items.splice(i, 1); }

  createIterator(): Iterator<Song> {
    return new PlaylistIterator(this);
  }
}